--- binutils-2.15.94.0.2.2/bfd/opncls.c~	2008-10-14 21:22:06.000000000 +0300
+++ binutils-2.15.94.0.2.2/bfd/opncls.c	2008-10-14 21:24:48.052157688 +0300
@@ -1141,6 +1141,10 @@
     }
   BFD_ASSERT (strlen (dir) != 0);
 
+  /* PLD: check the possible absolute debuginfo path (e.g. /tmp/.../usr/lib/debug/... */
+  if (separate_debug_file_exists (basename, crc32))
+    return basename;
+
   /* Strip off filename part.  */
   for (i = strlen (dir) - 1; i >= 0; i--)
     if (IS_DIR_SEPARATOR (dir[i]))
@@ -1345,6 +1349,15 @@
   return find_separate_debug_file (abfd, dir);
 }
 
+static char const*
+relaxed_filename (char const* filename)
+{
+  char const* relaxed = strstr (filename, "/usr/lib/debug/");
+  if ( relaxed != NULL )
+    return relaxed;
+  return filename;
+}
+
 /*
 FUNCTION
 	bfd_create_gnu_debuglink_section
@@ -1376,8 +1389,7 @@
       return NULL;
     }
 
-  /* Strip off any path components in filename.  */
-  filename = lbasename (filename);
+  filename = relaxed_filename (filename);
 
   sect = bfd_get_section_by_name (abfd, GNU_DEBUGLINK);
   if (sect)
@@ -1463,9 +1475,7 @@
     crc32 = bfd_calc_gnu_debuglink_crc32 (crc32, buffer, count);
   fclose (handle);
 
-  /* Strip off any path components in filename,
-     now that we no longer need them.  */
-  filename = lbasename (filename);
+  filename = relaxed_filename (filename);
 
   filelen = strlen (filename);
   debuglink_size = filelen + 1;
