--- binutils-2.16.90.0.3/bfd/elflink.c.orig	2005-05-12 00:27:10.731646000 +0200
+++ binutils-2.16.90.0.3/bfd/elflink.c	2005-05-12 23:05:36.202227648 +0200
@@ -7107,8 +7107,6 @@
 			    (_("`%s' referenced in section `%A' of %B: "
 			       "defined in discarded section `%A' of %B"),
 			     o, input_bfd, sec, sec->owner, sym_name);
-			  bfd_set_error (bfd_error_bad_value);
-			  return FALSE;
 			}
 
 		      /* Try to do the best we can to support buggy old
